core/server/common/src/main/java/alluxio/util/ParallelZipUtils.java

```diff
@@ -0,0 +1,180 @@
+import org.apache.commons.compress.archivers.zip.ZipArchiveOutputStream;
+import org.apache.commons.compress.archivers.zip.ZipFile;
+import org.apache.commons.compress.parallel.InputStreamSupplier;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.input.NullInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Enumeration;
+import java.util.concurrent.CompletionService;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.ExecutorService;
+import java.util.stream.Stream;
+
+/**
+ * Utility methods for working with parallel zip archives.
+ */
+public class ParallelZipUtils {
+  private static final Logger LOG = LoggerFactory.getLogger(ParallelZipUtils.class);
+
+  /**
+   * Creates a zipped archive from the given path in parallel, streaming the data
+   * to the give output stream.
+   *
+   * @param dirPath
+   * @param outputStream
+   * @param poolSize
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  public static void compress(Path dirPath, OutputStream outputStream, int poolSize)
+      throws IOException, InterruptedException {
+    LOG.info("compress in parallel for path {}", dirPath);
+    ExecutorService executor = ExecutorServiceFactories.fixedThreadPool(
+        "parallel-zip-compress-pool", poolSize).create();
+    ParallelScatterZipCreator parallelScatterZipCreator = new ParallelScatterZipCreator(executor);
+    ZipArchiveOutputStream zipArchiveOutputStream = new ZipArchiveOutputStream(outputStream);
+
+    try {
+      try (final Stream<Path> stream = Files.walk(dirPath)) {
+        for (Path subPath : stream.collect(toList())) {
+          if (Thread.interrupted()) {
+            throw new InterruptedException();
+          }
+
+          File file = subPath.toFile();
+
+          final InputStreamSupplier inputStreamSupplier = () -> {
+            try {
+              if (file.exists() && file.isFile()) {
+                return new FileInputStream(file);
+              } else {
+                return new NullInputStream(0);
+              }
+            } catch (FileNotFoundException e) {
+              LOG.warn("Can't find file when parallel zip, path = {}", subPath);
+              return new NullInputStream(0);
+            }
+          };
+
+          String entryName = dirPath.relativize(subPath).toString();
+          if (file.isDirectory()) {
+            entryName += File.separator;
+          }
+
+          ZipArchiveEntry zipArchiveEntry = new ZipArchiveEntry(entryName);
+          zipArchiveEntry.setMethod(ZipArchiveEntry.DEFLATED);
```

Comment:
I was thinking maybe its useful to have a property key to set the compression type, I suppose some people may prefer to use compression options in rocksDB, then disable compression here?

Reply:
Just curious, do you know approximately how the rates of compression compare between the different compression types? Also the overall CPU usage, like even if it is parallel is it using less or more CPU overall? I also see that the original Gzip compression streams has different levels of compression (0-9) I wonder how important these are.

Reply:
Apache common compress currently only supports parallel zip compression. 

I added a new option that can set the compression level, if the user does not want to compress, it can be set to 0.